sentry-tracing/src/layer.rs

```diff
@@ -102,9 +287,168 @@ impl<S: Subscriber> Layer<S> for SentryLayer {
+            }
+
+            let span = (self.span_mapper)(&span, None, attrs);
+            extensions.insert(Trace::new(span));
+        }
+    }
+
+    /// From the tracing-subscriber implementation of span timings,
+    /// keep track of when the span was last entered
+    fn on_enter(&self, id: &span::Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        if let Some(timings) = extensions.get_mut::<Trace>() {
+            let now = Instant::now();
+            timings.idle += (now - timings.last).as_nanos() as u64;
+            timings.last = now;
+        }
+    }
+
+    /// From the tracing-subscriber implementation of span timings,
+    /// keep track of when the span was last exited
+    fn on_exit(&self, id: &span::Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        if let Some(timings) = extensions.get_mut::<Trace>() {
+            let now = Instant::now();
+            timings.busy += (now - timings.last).as_nanos() as u64;
+            timings.last = now;
+            timings.last_sys = SystemTime::now();
+        }
+    }
+
+    /// When a span gets closed, if it has a trace extension either
+    /// attach it to a parent span or submit it as a Transaction if
+    /// it is a root of the span tree
+    fn on_close(&self, id: span::Id, ctx: Context<'_, S>) {
+        let span = ctx.span(&id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut trace = match extensions.remove::<Trace>() {
+            Some(trace) => trace,
+            None => return,
+        };
+
+        // Construct the timing data and call the on_close hook
+        trace.idle += (Instant::now() - trace.last).as_nanos() as u64;
+
+        let timings = Timings {
+            start_time: trace.first,
+            end_time: trace.last_sys,
+            idle: trace.idle,
+            busy: trace.busy,
+        };
+
+        (self.span_on_close)(&mut trace.span, timings);
+
+        // Traverse the parents of this span to attach to the nearest one
+        // that has tracing data (spans ignored by the span_filter do not)
+        for parent in span.parent().into_iter().flat_map(|span| span.scope()) {
+            let mut extensions = parent.extensions_mut();
+            if let Some(parent) = extensions.get_mut::<Trace>() {
+                parent.spans.extend(trace.spans);
+
+                trace.span.parent_span_id = Some(parent.span.span_id);
+                parent.spans.push(trace.span);
+                return;
+            }
+        }
+
+        // If no parent was found, consider this span a
+        // transaction root and submit it to Sentry
+        let span = &span;
+        Hub::with_active(move |hub| {
+            let client = hub.client().unwrap();
```

Comment:
please match on this rather than `unwrap`

Reply:
Ah yes, true. Though being defensive with regard to panics is a good idea in general.